rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,593 @@
+// Utility for logging global region manager state (in stable memory).
+// For sanity-checking during testing and for future trouble-shooting.
+// (Perhaps we can keep this here, and just not commit to it when exposing final API?)
+#[ic_mem_fn]
+pub unsafe fn region_meta_loglines<M: Memory>(_mem: &mut M) {
+    let log_id = NEXT_REGION_LOG_ID;
+    NEXT_REGION_LOG_ID += 1;
+    println!(50, "# regionMetaLogLines {{");
+    println!(50, " log_id = {};", log_id);
+    println!(
+        50,
+        " total_allocated_blocks = {};",
+        meta_data::total_allocated_blocks::get()
+    );
+    println!(
+        50,
+        " total_allocated_regions = {};",
+        meta_data::total_allocated_regions::get()
+    );
+    println!(50, "}}");
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_id<M: Memory>(_mem: &mut M, r: Value) -> u32 {
+    let r = r.as_region();
+    (*r).id.into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_size<M: Memory>(_mem: &mut M, r: Value) -> u64 {
+    let r = r.as_region();
+    (*r).page_count.into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_grow<M: Memory>(mem: &mut M, r: Value, new_pages: u64) -> u64 {
+    use meta_data::size::{required_pages, PAGES_IN_BLOCK};
+    let r = r.as_region();
+    let new_pages_ = new_pages as u32;
+    let old_page_count = (*r).page_count;
+    let old_block_count = (old_page_count + (PAGES_IN_BLOCK - 1)) / PAGES_IN_BLOCK;
+    let new_block_count = (old_page_count + new_pages_ + (PAGES_IN_BLOCK - 1)) / PAGES_IN_BLOCK;
+    let inc_block_count = new_block_count - old_block_count;
+
+    println!(
+        80,
+        "begin region_grow id={} page_count={} new_pages={}",
+        (*r).id,
+        old_page_count,
+        new_pages
+    );
+
+    // Update the total number of allocated blocks.
+    let old_total_blocks = {
+        let c = meta_data::total_allocated_blocks::get();
+        let c_ = c + inc_block_count as u64;
+        // to do -- assert c_ is less than meta_data::max::BLOCKS
+        meta_data::total_allocated_blocks::set(c_);
+        c
+    };
+
+    // Actually grow stable memory with more pages
+    // (but only if needed, by first checking if someone else did it already).
+    {
+        let have = crate::ic0_stable::nicer::size();
+        let need = required_pages();
+        if have < need {
+            let diff = need - have;
+            crate::ic0_stable::nicer::grow(diff);
+        }
+    }
+
+    // Update this region's page count, in both places where we record it (heap object, region table).
+    {
+        let r_id = RegionId::from_id((*r).id);
+        let c = meta_data::region_table::get(&r_id);
+
+        // Region table agrees with heap object's field.
+        assert_eq!(c, Some(RegionSizeInPages(old_page_count.into())));
+
+        // Increase both:
+        (*r).page_count += new_pages_;
+        meta_data::region_table::set(&r_id, Some(RegionSizeInPages((*r).page_count.into())));
+    }
+
+    let new_vec_pages = alloc_blob(mem, Bytes(new_block_count * 2));
```

Comment:
Here you correctly allocate new_block_count * 2 bytes, but don't ask for twice the bytes above.